src/ngraph/runtime/gpu/cudnn_emitter.cpp

```diff
@@ -0,0 +1,168 @@
+        get_input_desc = [dimensions]() {
+            float* x = new float();
+
+            cudnnTensorDescriptor_t desc;
+            cudnnCreateTensorDescriptor(&desc);
+            cudnnSetTensorNdDescriptor(desc,
+                                       CUDNN_DATA_FLOAT,
+                                       dimensions.size(),
+                                       dimensions.data(),
+                                       cudnn_util::compute_strides(dimensions).data());
+            return desc;
+        };
+
+        // mark reduced axes of input tensor for output tensor descriptor
+        for (auto const& idx_dim : reduction_axes)
+        {
+            dimensions[idx_dim] = 1;
+        }
+
+        get_output_desc = [dimensions]() {
+            cudnnTensorDescriptor_t desc;
+            cudnnCreateTensorDescriptor(&desc);
+            cudnnSetTensorNdDescriptor(desc,
+                                       CUDNN_DATA_FLOAT,
+                                       dimensions.size(),
+                                       dimensions.data(),
+                                       cudnn_util::compute_strides(dimensions).data());
+            return desc;
+        };
+    }
+    // emit sum reduce operation
+    auto* reduce = new gpu::primitive{
+        [ctx, reduce_op, get_input_desc, get_output_desc](void** inputs, void** outputs) {
+            auto input_desc = get_input_desc();
+            auto output_desc = get_output_desc();
+            cudnnReduceTensorDescriptor_t reduceTensorDesc;
+            cudnnCreateReduceTensorDescriptor(&reduceTensorDesc);
+            cudnnSetReduceTensorDescriptor(reduceTensorDesc,
+                                           reduce_op,
+                                           CUDNN_DATA_FLOAT,
+                                           CUDNN_NOT_PROPAGATE_NAN,
+                                           CUDNN_REDUCE_TENSOR_NO_INDICES,
+                                           CUDNN_32BIT_INDICES);
+            size_t workspace_size = 0;
+            cudnnGetReductionWorkspaceSize(
+                *ctx->cudnn_handle, reduceTensorDesc, input_desc, output_desc, &workspace_size);
+            auto workspace_ptr = create_gpu_buffer(workspace_size);
+            float alpha = 1.0, beta = 0.0;
+            cudnnReduceTensor(*ctx->cudnn_handle,
+                              reduceTensorDesc,
+                              nullptr,
+                              0,
+                              workspace_ptr,
+                              workspace_size,
+                              &alpha,
+                              input_desc,
+                              inputs[0],
+                              &beta,
+                              output_desc,
+                              outputs[0]);
+            free_gpu_buffer(workspace_ptr);
+        }};
+
+    return this->m_primitive_emitter->insert(reduce);
+}
+
+std::vector<int> cudnn_util::compute_strides(const std::vector<int>& dim)
```

Comment:
This must exist somewhere else, but I haven't looked. Ask Scott

Reply:
Thanks, I have added a cudnn_util functions to use the ngraph::row_major_strides function. But actual in this case, I need to convert the shape from td::vector<size_t> to std::vector<int> and then do the stride calculation, as both the shape and strides are required to be arrays of ints when passed to cudnn.